From 44a30a2b675839b0312ead7c93a7e94ff4c226b0 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Thu, 21 Mar 2024 14:16:26 +0100
Subject: [PATCH 0876/1695] plat-stm32mp2: add SCMI power domain set-up

Enable compilation of mod_power_domain (SCP generic),
mod_scmi_power_domain (SCP generic) and
new mod_stm32_pd (specific to stm32mp2 platform) in SCP firmware.

Power domains are described in stm32_scmi_pd array.
A power domain control one clock and one regulator.

Give to SCP theses information through an scmi_pd structure.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f47279ab88f3442cc2fc06bc37130d3d452487a6 plat-stm32mp2: add SCMI power domain set-up]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I218f483a34946652cb8371b04aae0b947a7bea42
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/373696
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     | 72 +++++++++++++++++++
 core/lib/scmi-server/conf-optee-stm32mp2.mk   |  7 ++
 .../include/scmi_agent_configuration.h        | 16 +++++
 core/lib/scmi-server/sub-optee-stm32mp2.mk    |  5 ++
 4 files changed, 100 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index 104f3beb5..0d5f3109e 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -76,6 +76,18 @@ struct stm32_scmi_perfd {
 	const char *name;
 };
 
+/*
+ * struct stm32_scmi_pd - Data for the exposed power domains
+ * @name: Power domain name exposed to the channel
+ * @clock_id: Clock identifier in RCC clock driver
+ * @regu_name: Regulator node name in OP-TEE secure FDT
+ */
+struct stm32_scmi_pd {
+	const char *name;
+	unsigned long clock_id;
+	const char *regu_name;
+};
+
 #define CLOCK_CELL(_scmi_id, _id, _name, _init_enabled) \
 	[_scmi_id] = { \
 		.clock_id = _id, \
@@ -106,6 +118,13 @@ struct stm32_scmi_perfd {
 		.name = (_name), \
 	}
 
+#define PD_CELL(_scmi_id, _name, _clock_id, _regu_name) \
+	[_scmi_id] = { \
+		.name = (_name), \
+		.clock_id = (_clock_id), \
+		.regu_name = (_regu_name), \
+	}
+
 struct channel_resources {
 	struct stm32_scmi_clk *clock;
 	size_t clock_count;
@@ -113,6 +132,8 @@ struct channel_resources {
 	size_t rd_count;
 	struct stm32_scmi_perfd *perfd;
 	size_t perfd_count;
+	struct stm32_scmi_pd *pd;
+	size_t pd_count;
 };
 
 #ifdef CFG_STM32MP21
@@ -441,12 +462,22 @@ static struct stm32_scmi_rd stm32_scmi_reset[] = {
 #endif
 };
 
+#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
+static struct stm32_scmi_pd stm32_scmi_pd[] = {
+	PD_CELL(PD_SCMI_GPU, "gpu", PLL3_CK, "vddgpu"),
+};
+#endif
+
 static const struct channel_resources scmi_channel[] = {
 	[0] = {
 		.clock = stm32_scmi_clock,
 		.clock_count = ARRAY_SIZE(stm32_scmi_clock),
 		.rd = stm32_scmi_reset,
 		.rd_count = ARRAY_SIZE(stm32_scmi_reset),
+#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
+		.pd = stm32_scmi_pd,
+		.pd_count = ARRAY_SIZE(stm32_scmi_pd),
+#endif
 	},
 };
 
@@ -694,6 +725,46 @@ static TEE_Result scmi_scpfw_cfg_init(void)
 				};
 			}
 		}
+
+#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
+		channel_cfg->pd_count = ARRAY_SIZE(stm32_scmi_pd);
+		if (channel_cfg->pd_count) {
+			channel_cfg->pd = calloc(channel_cfg->pd_count,
+						 sizeof(*channel_cfg->pd));
+			assert(channel_cfg->pd);
+
+			for (j = 0; j < channel_cfg->pd_count; j++) {
+				struct clk *clk = stm32mp_rcc_clock_id_to_clk(
+					stm32_scmi_pd[j].clock_id);
+				struct regulator *regu = NULL;
+
+				if (stm32_scmi_pd[j].regu_name != NULL) {
+					regu = regulator_get_by_name(
+						stm32_scmi_pd[j].regu_name);
+
+					if (!regu) {
+						EMSG("Regulator %s not found "
+						     "for %s power domain",
+						     stm32_scmi_pd[j].regu_name,
+						     stm32_scmi_pd[j].name);
+						panic();
+					}
+				}
+
+				if (!clk)
+					DMSG("Clock ID %ld not found for %s "
+					     "power domain",
+					     stm32_scmi_pd[j].clock_id,
+					     stm32_scmi_pd[j].name);
+
+				channel_cfg->pd[j] = (struct scmi_pd){
+					.name = stm32_scmi_pd[j].name,
+					.clk = clk,
+					.regu = regu,
+				};
+			}
+		}
+#endif
 	}
 
 	/* DVFS will be populated from cpu_opp driver */
@@ -739,6 +810,7 @@ void scmi_scpfw_release_configuration(void)
 			}
 
 			free(channel_cfg->perfd);
+			free(channel_cfg->pd);
 		}
 
 		free(agent_cfg->channel_config);
diff --git a/core/lib/scmi-server/conf-optee-stm32mp2.mk b/core/lib/scmi-server/conf-optee-stm32mp2.mk
index 371b308e3..76bfeaeec 100644
--- a/core/lib/scmi-server/conf-optee-stm32mp2.mk
+++ b/core/lib/scmi-server/conf-optee-stm32mp2.mk
@@ -14,6 +14,13 @@ $(call force,CFG_SCPFW_MOD_SCMI_CLOCK,y)
 $(call force,CFG_SCPFW_MOD_SCMI_PERF,y)
 $(call force,CFG_SCPFW_MOD_SCMI_RESET_DOMAIN,y)
 $(call force,CFG_SCPFW_MOD_SCMI_VOLTAGE_DOMAIN,y)
+
+ifeq ($(call cfg-one-enabled, CFG_STM32MP23 CFG_STM32MP25),y)
+$(call force,CFG_SCPFW_MOD_POWER_DOMAIN,y)
+$(call force,CFG_SCPFW_MOD_SCMI_POWER_DOMAIN,y)
+$(call force,CFG_SCPFW_MOD_STM32_PD,y)
+endif
+
 $(call force,CFG_SCPFW_MOD_VOLTAGE_DOMAIN,y)
 $(call force,CFG_SCPFW_SCMI_PERF_PROTOCOL_OPS,y)
 
diff --git a/core/lib/scmi-server/include/scmi_agent_configuration.h b/core/lib/scmi-server/include/scmi_agent_configuration.h
index 32a95260a..b9dc94efe 100644
--- a/core/lib/scmi-server/include/scmi_agent_configuration.h
+++ b/core/lib/scmi-server/include/scmi_agent_configuration.h
@@ -71,6 +71,18 @@ struct scmi_perfd {
 	struct regulator *regulator;
 };
 
+/*
+ * struct scmi_pd - Description of a power domain resource
+ * @name: Domain name
+ * @clk: Clock controlled by the power domain or NULL
+ * @regu: Voltage regulator controlled by the power domain or NULL
+ */
+struct scmi_pd {
+	const char *name;
+	struct clk *clk;
+	struct regulator *regu;
+};
+
 /*
  * struct scpfw_channel_config - SCMI channel resources
  * @name: Channel name
@@ -83,6 +95,8 @@ struct scmi_perfd {
  * @voltd_count: Number of cells of @voltd
  * @perfd: Description of the DVFS/perf exposed on the channel
  * @perfd_count: Number of cells of @perfd
+ * @pd: Description of the power domain exposed on the channel
+ * @pd_count: Number of cells of @pd
  */
 struct scpfw_channel_config {
 	const char *name;
@@ -95,6 +109,8 @@ struct scpfw_channel_config {
 	size_t voltd_count;
 	struct scmi_perfd *perfd;
 	size_t perfd_count;
+	struct scmi_pd *pd;
+	size_t pd_count;
 };
 
 /*
diff --git a/core/lib/scmi-server/sub-optee-stm32mp2.mk b/core/lib/scmi-server/sub-optee-stm32mp2.mk
index f275178ce..bba6e7671 100644
--- a/core/lib/scmi-server/sub-optee-stm32mp2.mk
+++ b/core/lib/scmi-server/sub-optee-stm32mp2.mk
@@ -5,3 +5,8 @@ srcs-y += $(scpfw-path)/product/optee-stm32mp2/fw/config_all.c
 ifeq ($(CFG_SCPFW_MOD_PSU_OPTEE_REGULATOR),y)
 $(eval $(call scpfw-embed-product-module,psu_optee_regulator))
 endif
+
+ifeq ($(CFG_SCPFW_MOD_STM32_PD),y)
+$(eval $(call scpfw-embed-product-module,stm32_pd))
+endif
+
-- 
2.47.2

